### DISCUSSION TITLE
Vercel Postgres: update multiple rows in one query

### DISCUSSION THREAD

**Discussion by dmackca** [vercel/community#3682] [Feature Question]
### Question

I have a serverless function that may need to update several rows in a table. I'd like to do this all in one query:

As an example, an episode tracker for podcasts:
```sql
UPDATE subscriptions AS s
      SET
          latestepisode = u.new_latestepisode
      FROM (VALUES
          (1, 'S03E04'),
          (4, 'S07E01')
      ) AS u(id, new_latestepisode)
      WHERE s.id = u.id;
 ```
That query works fine if I put it into the Vercel Postgres UI dashboard thing, but I can't get it to work with the `sql` template literal tag.
I can't keep all of the parameters tokenized if I've got a dynamic number of them. (Unless somebody knows how?)

I tried building the dynamic part as a string, like:
```sql
await sql`
      UPDATE subscriptions AS s
      SET
          latestepisode = u.new_latestepisode
      FROM (VALUES
          ${updateRows}
      ) AS u(id, new_latestepisode)
      WHERE s.id = u.id;
`;
```
But I got an error:

> NeonDbError: db error: ERROR: syntax error at or near "$1"

Since the docs say the sql tag will convert raw SQL into parameterized SQL, I also tried building the whole query as a string:
```sql
await sql`${query}`;
```

But I get the same error.

Is there any way to do this with the Vercel Postgres SDK `sql` tag, or do I need to use a more advanced query builder?

**Comment by dmarkbreiter** [comment]
@dmackca, I'm running into the same issue when using expressions for field names. I've switched to Prisma which is kind of overkill for my use case but I couldn't find documentation/solution for this error.

**Reply by dmackca** [reply]
I'll probably end up doing the same thing, but I was hoping the Vercel SQL SDK had a function similar to pg-promise where you can provide a query with tokens like $<myVar> and then pass in the values for those variables separately.

Because, yeah, Prisma does seem like overkill for my use case 😆

**Reply by dmackca** [reply]
...and I suppose, for my use case, I could get away with just doing multiple queries from the same connection (using the `client.sql` syntax) but that feels a little dirty for what I'm trying to do.

edit: Maybe the `client` object has the same API as the node-postgres client, so we could do something like this;
https://node-postgres.com/features/queries#parameterized-query

I'll try that this weekend and report back.

**Comment by zackkrida** [comment]
Here's the code where Vercel turns the tagged template literal into a parameterized query:

https://raw.githubusercontent.com/vercel/storage/main/packages/postgres/src/sql-template.ts

For example:

```js

let breed = 'golden retriever'
sqlTemplate`SELECT * from dogs where breed = ${breed}`
// Produces: [
//   'SELECT * from dogs where breed = $1',
//   [ 'golden retriever' ]
// ]
```

An example with an array:

```js
let teamName = 'test 123'
let salesReps = [104,105,1134]

sqlTemplate`UPDATE sales_rep SET team = ${teamName} WHERE sales_rep_id = ANY (${salesReps});`

// Produces: [
//   'UPDATE sales_rep SET team = $1 WHERE sales_rep_id = ANY ($2);',
//   [ 'test 123', [ 104, 105, 1134 ] ]
// ]
```

Hopefully this helps you fix your query 😄

**Reply by dmackca** [reply]
Unfortunately that won't work for my query, because it's being built with a dynamic number of parameters (therein lies the problem!)

**Comment by dmackca** [comment]
I got my query working with the @vercel/postgres SDK but it's not pretty 😆 
You can indeed write a parameterized query as per the node-postgres documentation, but there [seems to be a bug](https://github.com/brianc/node-postgres/issues/2724) where you have to explicitly cast numeric types, otherwise they get parsed as a string.

So in the end I built a string like `"($1, $2), ($3, $4)"` (and so on) and an array of query parameters like `[1, "S01E04', 7, "S03E01"]` and ran the query with the `client.query` method:

```js
const client = await sql.connect();
await client.query(query, queryParameters);
client.release();
```

That should work, but because of the (bug?) mentioned above, I had to cast the data types in the query, changing my string to:
```
"($1::int, $2::text), ($3::int, $4::text)"
```

I'd be curious to hear if @dmarkbreiter had an easier time of it with Prisma. Let me know!

**Comment by dmackca** [comment]
Sort of a kafka-esque epilogue to this story; my hobby project exhausted Vercel Postgres' compute time limits before I could even launch it, so I just ditched it. I'm now using Supabase for my DB, and their Javascript API has an easy way to update multiple rows in one query 👍
